crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,320 @@
+            }
+            _ => Vertex::ASTNode(ast_node),
+        }
+    }
+}
+
+impl<'a> From<&'a Expression<'a>> for Vertex<'a> {
+    fn from(expr: &'a Expression<'a>) -> Self {
+        // FIXME: We just get rid of all parentheses here, but we shouldn't do that...
+
+        // NOTE: When string literal / template literal is added, add to as_constant_string
+        match &expr.get_inner_expression() {
+            Expression::ObjectExpression(objexpr) => Vertex::ObjectLiteral(objexpr),
+            Expression::JSXElement(element) => {
+                Vertex::JSXElement(JSXElementVertex { ast_node: None, element }.into())
+            }
+            _ => Vertex::Expression(expr),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClassVertex<'a> {
+    ast_node: Option<AstNode<'a>>,
+    pub class: &'a Class<'a>,
+}
+
+impl<'a> Typename for ClassVertex<'a> {
+    fn typename(&self) -> &'static str {
+        if self.ast_node.is_some() { "ClassAST" } else { "Class" }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClassMethodVertex<'a> {
+    pub method: &'a MethodDefinition<'a>,
+    pub is_abstract: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct ClassPropertyVertex<'a> {
+    pub property: &'a PropertyDefinition<'a>,
+    pub is_abstract: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct ImportVertex<'a> {
+    ast_node: Option<AstNode<'a>>,
+    pub import: &'a ImportDeclaration<'a>,
+}
+
+impl<'a> Typename for ImportVertex<'a> {
+    fn typename(&self) -> &'static str {
+        if self.ast_node.is_some() { "ImportAST" } else { "Import" }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InterfaceVertex<'a> {
+    ast_node: Option<AstNode<'a>>,
+    pub interface: &'a TSInterfaceDeclaration<'a>,
+}
+
+impl<'a> Typename for InterfaceVertex<'a> {
+    fn typename(&self) -> &'static str {
+        if self.ast_node.is_some() { "InterfaceAST" } else { "Interface" }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum InterfaceExtendVertex<'a> {
+    Identifier(&'a IdentifierReference),
+    MemberExpression(&'a MemberExpression<'a>),
+}
+
+#[derive(Debug, Clone)]
+pub struct JSXElementVertex<'a> {
+    pub ast_node: Option<AstNode<'a>>,
+    pub element: &'a JSXElement<'a>,
+}
+
+impl<'a> Typename for JSXElementVertex<'a> {
+    fn typename(&self) -> &'static str {
+        if self.ast_node.is_some() { "JSXElementAST" } else { "JSXElement" }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ReturnStatementVertex<'a> {
+    ast_node: Option<AstNode<'a>>,
+    pub return_statement: &'a ReturnStatement<'a>,
+}
+
+#[derive(Debug, Clone)]
+pub struct TypeAnnotationVertex<'a> {
+    ast_node: Option<AstNode<'a>>,
+    pub type_annotation: &'a TSTypeAnnotation<'a>,
+}
+
+impl<'a> Typename for TypeAnnotationVertex<'a> {
+    fn typename(&self) -> &'static str {
+        if self.ast_node.is_some() { "TypeAnnotationAST" } else { "TypeAnnotation" }
+    }
+}
+
+#[derive(Debug, Clone)]
```

Comment:
```suggestion
#[non_exhaustive]
#[derive(Debug, Clone)]
```
Otherwise if you add any fields to this struct (public or private), it's a major breaking change. Probably a good idea to add to `Vertex` and elsewhere as well!

Reply:
Why would this matter? since we don't expose the adapter's vertex externally, do we?

Reply:
I just realized we do, good catch, I just won't expose the vertex publicly and this shouldn't be a problem, I think.

Reply:
[dbb13eb](https://github.com/web-infra-dev/oxc/pull/599/commits/dbb13eb076be90d680198bf0d2735e20ad0b6ee9)

Reply:
It doesn't cost anything to add, and if it somehow unexpectedly gets exposed, it won't cause a big problem when it changes. It's harder to show that it's not exposed externally than to make it non-exhaustive.

Reply:
Good thought, I'll add it

Reply:
Is there any way to say that this shouldn't be exposed outside of this crate instead? since in many cases I rely on the ability to fully pattern match on Vertex, which I can't do with this change?

Reply:
Ah whatever, I just made it non-exhaustive